crates/levm/src/vm.rs

```diff
@@ -518,70 +728,148 @@ impl VM {
-                    let mut value_bytes = [0u8; 32];
+                    let mut value_bytes = [0u8; WORD_SIZE];
                     value.to_big_endian(&mut value_bytes);
 
                     current_call_frame.memory.store_bytes(offset, &value_bytes);
+                    tx_env.consumed_gas += gas_cost
                 }
                 Opcode::MSTORE8 => {
-                    // spend_gas(3);
                     let offset = current_call_frame.stack.pop().unwrap().try_into().unwrap();
+                    let memory_expansion_cost =
+                        current_call_frame.memory.expansion_cost(offset + 1);
+                    let gas_cost = gas_cost::MSTORE8_STATIC + memory_expansion_cost;
+                    if tx_env.consumed_gas + gas_cost > tx_env.gas_limit {
+                        break; // should revert the tx
+                    }
+
                     let value = current_call_frame.stack.pop().unwrap();
-                    let mut value_bytes = [0u8; 32];
+                    let mut value_bytes = [0u8; WORD_SIZE];
                     value.to_big_endian(&mut value_bytes);
 
                     current_call_frame
                         .memory
-                        .store_bytes(offset, value_bytes[31..32].as_ref());
+                        .store_bytes(offset, value_bytes[WORD_SIZE - 1..WORD_SIZE].as_ref());
+                    tx_env.consumed_gas += gas_cost
                 }
                 Opcode::MSIZE => {
-                    // spend_gas(2);
+                    if tx_env.consumed_gas + gas_cost::MSIZE > tx_env.gas_limit {
+                        break; // should revert the tx
+                    }
                     current_call_frame
                         .stack
                         .push(current_call_frame.memory.size());
+                    tx_env.consumed_gas += gas_cost::MSIZE
+                }
+                Opcode::GAS => {
+                    if tx_env.consumed_gas + gas_cost::GAS > tx_env.gas_limit {
+                        break; // should revert the tx
+                    }
+                    let remaining_gas = tx_env.gas_limit - tx_env.consumed_gas - gas_cost::GAS;
+                    current_call_frame.stack.push(remaining_gas.into());
+                    tx_env.consumed_gas += gas_cost::GAS
                 }
                 Opcode::MCOPY => {
-                    // spend_gas(3) + dynamic gas
                     let dest_offset = current_call_frame.stack.pop().unwrap().try_into().unwrap();
-                    let src_offset = current_call_frame.stack.pop().unwrap().try_into().unwrap();
-                    let size = current_call_frame.stack.pop().unwrap().try_into().unwrap();
+                    let src_offset: usize =
+                        current_call_frame.stack.pop().unwrap().try_into().unwrap();
+                    let size: usize = current_call_frame.stack.pop().unwrap().try_into().unwrap();
                     if size == 0 {
```

Comment:
Here if size is 0 I think we aren't consuming any gas, shouldn't we atleast consume `gas_cost::MCOPY_STATIC`?

Reply:
Yes, just moved that gas consumption before the `if` statement.